lib/client/ClientSocket.js

```diff
@@ -0,0 +1,92 @@
+/*
+  Copyright © 2018 Andrew Powell
+
+  This Source Code Form is subject to the terms of the Mozilla Public
+  License, v. 2.0. If a copy of the MPL was not distributed with this
+  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+  The above copyright notice and this permission notice shall be
+  included in all copies or substantial portions of this Source Code Form.
+*/
+const { error, refresh, warn } = require('./log');
+
+// ignore 1008 (HTTP 400 equivalent) and 1011 (HTTP 500 equivalent)
+const ignoreCodes = [1008, 1011];
+const maxAttempts = 10;
+
+class ClientSocket {
+  constructor(options, ...args) {
+    this.args = args;
+    this.attempts = 0;
+    this.eventHandlers = [];
+    this.options = options;
+    this.retrying = false;
+
+    this.connect();
+  }
+
+  addEventListener(...args) {
```

Comment:
cleaver!! You created a class with almost the same methods that a `WebSocket` instance has!!!